python/ray/client_builder.py

```diff
@@ -63,6 +67,33 @@ class ClientBuilder:
     def __init__(self, address: Optional[str]) -> None:
         self.address = address
         self._job_config = JobConfig()
+        self._init_args_dict = {}
+        self._catch_all_dict = {}
+
+        # Check environment variables for default values
+        namespace_env_var = os.environ.get(RAY_NAMESPACE_ENVIRONMENT_VARIABLE)
+        if namespace_env_var:
+            self.namespace(namespace_env_var)
+
+        runtime_env_var = os.environ.get(RAY_RUNTIME_ENV_ENVIRONMENT_VARIABLE)
+        if runtime_env_var:
+            self.env(json.loads(runtime_env_var))
+
+        # Check address for values
+        if address:
+            # Add a dummy protocol to parse correctly
+            parsed_address = urlparse("dummyProto://" + address)
```

Comment:
why do we need to add dummyProto?

Reply:
If the address is something like "localhost:50005?query=thing" then urlparse will mix up `localhost` as the scheme (protocol)

```python
>>> urlparse("localhost:8000?a=b")
ParseResult(scheme='localhost', netloc='', path='8000', params='', query='a=b', fragment='')
>>> urlparse("hi://localhost:8000?a=b")
ParseResult(scheme='hi', netloc='localhost:8000', path='', params='', query='a=b', fragment='')
```

^ in the second line specifying the scheme makes sure the whole address lands in `netloc`

Reply:
but wait, this looks like an illegal address, because we only support
XXX://<string>
or
<string>
you can't do localhost:50005?query=thing. Actually you can't add `?query=thing` to any option
.